stash_engine/app/controllers/stash_engine/dashboard_controller.rb

```diff
@@ -38,5 +49,63 @@ def ajax_wait
         end
       end
     end
+
+    # methods below are private
+    private
+
+    def email_code
+      current_user.update(old_dryad_email: params[:email])
+      current_user.set_migration_token
+      StashEngine::MigrationMailer.migration_email(email: current_user.old_dryad_email,
+                                                   code: current_user.migration_token,
+                                                   url: auth_migrate_code_url).deliver_now
+    end
+
+    def validate_form_email
+      if params[:email].nil? || !params[:email][/^.+\@.+\..+$/]
+        flash.now[:info] = 'Please fill in a correct email address' if params[:commit]
+        render 'migrate_data'
+        return false
+      end
+      true
+    end
+
+    def validate_form_token_format
+      if params[:code].nil? || !params[:code][/^\d{6}$/]
+        flash.now[:info] = 'Please enter your 6-digit code to migrate your data' if params[:commit] == 'Migrate data'
```

Comment:
Just to cover the case in which the user enters an incorrect code, this should say "Please enter your *correct* 6-digit code..."

Reply:
changed